signer/src/bitcoin/packaging.rs

```diff
@@ -55,132 +77,428 @@ where
+    /// `true` if the item can be safely added to this bag.
+    fn is_compatible(&self, item: &T) -> bool {
+        self.votes_compatible(item)
+            && self.signatures_compatible(item)
+            && self.withdrawal_id_compatible(item)
+    }
+
+    /// Check if an item's votes are compatible with this bag.
+    ///
+    /// ## Parameters
+    /// - `item`: Item to check for vote compatibility
+    ///
+    /// ## Returns
+    /// `true` if the combined votes don't exceed the maximum allowed.
+    fn votes_compatible(&self, item: &T) -> bool {
+        let combined_votes = self.votes_bitmap | item.votes();
+        combined_votes.count_ones() <= self.config.max_votes_against
+    }
+
+    /// Check if an item's signature requirement is compatible with this bag.
+    ///
+    /// ## Parameters
+    /// - `item`: Item to check for signature compatibility
+    ///
+    /// ## Returns
+    /// `true` if adding the item wouldn't exceed the signature limit.
+    fn signatures_compatible(&self, item: &T) -> bool {
+        let sig = item.needs_signature() as u16;
+        self.items_needing_signatures + sig <= self.config.max_signatures
+    }
+
+    /// Check if an item's withdrawal ID is compatible with this bag.
+    ///
+    /// ## Parameters
+    /// - `item`: Item to check for withdrawal ID compatibility
+    ///
+    /// ## Returns
+    /// `true` if the item's withdrawal ID can fit in this bag's OP_RETURN.
+    fn withdrawal_id_compatible(&self, item: &T) -> bool {
+        let Some(id) = item.withdrawal_id() else {
+            return true;
+        };
+
+        matches!(self.can_add_withdrawal_id(id), Ok(true))
+    }
+
+    /// Calculate compatibility score between item and bag (smaller is better).
+    ///
+    /// The score is based on how different the vote patterns are (using XOR).
+    /// Lower scores indicate items with more similar voting patterns.
+    ///
+    /// ## Parameters
+    /// - `item`: Item to calculate compatibility score for
+    ///
+    /// ## Returns
+    /// A score where lower values indicate better compatibility.
+    fn compatibility_score(&self, item: &T) -> u32 {
+        // XOR measures how different the vote patterns are
+        (self.votes_bitmap ^ item.votes()).count_ones()
+    }
+
+    /// Check if adding a single withdrawal ID would exceed the OP_RETURN size
+    /// limit.
+    ///
+    /// ## Parameters
+    /// - `new_id`: Withdrawal ID to check
+    ///
+    /// ## Returns
+    /// - `Ok(true)` if the ID can be added
+    /// - `Ok(false)` if adding the ID would exceed size limits
+    /// - `Err(e)` on error
+    ///
+    /// ## Implementation Notes
+    /// This method simulates adding the new withdrawal ID to the bag's existing
+    /// IDs while maintaining sorted order. The [`BitmapSegmenter`] is then used
+    /// to estimate the size of the combined IDs, which requires sorted and
+    /// deduplicated IDs.
+    fn can_add_withdrawal_id(&self, new_id: u64) -> Result<bool, Error> {
```

Comment:
This doesn't need to return a `Result`

Reply:
Right, leftover, changed in [b02d783](https://github.com/stacks-network/sbtc/pull/1512/commits/b02d7837663c4cef165471cfb46cb1c4615b92ef)